sympy/vector/operators.py

```diff
@@ -252,21 +271,29 @@ def gradient(scalar_field, coord_sys=None, doit=True):
     """
     coord_sys = _get_coord_sys_from_expr(scalar_field, coord_sys)
 
-    if coord_sys is None:
+    if len(coord_sys) == 0:
         return Vector.zero
-    else:
+    elif len(coord_sys) == 1:
+        coord_sys = coord_sys.pop()
         from sympy.vector.functions import express
         h1, h2, h3 = coord_sys.lame_coefficients()
         i, j, k = coord_sys.base_vectors()
         x, y, z = coord_sys.base_scalars()
-        scalar_field = express(scalar_field, coord_sys, variables=True)
+        #scalar_field = express(scalar_field, coord_sys, variables=True)
         vx = Derivative(scalar_field, x) / h1
         vy = Derivative(scalar_field, y) / h2
         vz = Derivative(scalar_field, z) / h3
 
         if doit:
             return (vx * i + vy * j + vz * k).doit()
         return vx * i + vy * j + vz * k
+    else:
+        if isinstance(scalar_field, (Add, VectorAdd)):
+            return VectorAdd.fromiter(gradient(i) for i in scalar_field.args)
+        if isinstance(scalar_field, (Mul, VectorMul)):
+            s = _split_mul_args_wrt_coordsys(scalar_field)
+            return VectorAdd.fromiter(scalar_field / i * gradient(i) for i in s)
+        return Gradient(scalar_field)
```

Comment:
@szymag this is what I meant. One case for the addition, and another one for the multiplication. It looks a lot more readable.